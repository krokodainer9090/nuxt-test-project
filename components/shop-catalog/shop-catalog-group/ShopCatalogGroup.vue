<script>
// Components
import ShopCatalogGroupItem from './ShopCatalogGroupItem';

export default {
  name: 'ShopCatalogGroup',

  components: {
    ShopCatalogGroupItem
  },

  props: {
    group: {
      type: Object,
      default: _ => ({})
    }
  },

  computed: {
    groupItems () {
      return this.group?.items || [];
    }
  },

  methods: {
    getGroupItemKey (key) {
      return `shop-catalog-group-item-${key}`;
    },

    onClickItem (item) {
      this.$emit('on-click-item', item);
    }
  }
};
</script>

<template>
  <div class="shop-catalog-group">
    <h3 class="shop-catalog-group__title">{{ group.G }}</h3>
    <div class="shop-catalog-group__content container">
      <div class="row">
        <ShopCatalogGroupItem
          v-for="(item, index) in groupItems"
          :key="getGroupItemKey(item.id || index)"
          :item="item"
          class="shop-catalog-group__group-item col-12 col-xl-6"
          @on-click-item="onClickItem"
        />
      </div>
    </div>
  </div>
</template>

<style lang="scss">
@import "../../../styles/components/shop-catalog/shop-catalog-group/shop-catalog-group.scss";
</style>
