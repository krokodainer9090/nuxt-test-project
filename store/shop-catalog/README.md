Будет хранить всю информацию о каталоге в сторе.

С помощью action'а getShopCatalog получаем изначально состояние каталога, а с помощью
action'а fetchCatalogWithCurrency обновляем значение валюты и заново задаем состояние, на основе полученных данных.

Также fetchCatalogWithCurrency обновляет все цены (поле price) в корзине пользователя.
