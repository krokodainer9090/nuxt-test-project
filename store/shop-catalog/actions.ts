// Json Data
import data from '../../static/data/data.json';
import names from '../../static/data/names.json';

// Types
import {
  ShopCatalogParams,
  ShopCatalogState,
  ShopCatalog,
  ShopCatalogGroup,
  ShopCatalogItem, ShopCatalogClient
} from '~/store/shop-catalog/types';

// Utilities
import { getRandomNumberBetween } from '~/assets/ts/helpers';

// Vuex
import { ActionTree, ActionContext } from 'vuex';

const actions: ActionTree<ShopCatalogState, any> = {
  async getShopCatalog ({ commit, rootGetters }: ActionContext<ShopCatalogState, any>,
    { goodsValue, namesValue }: ShopCatalogParams = { goodsValue: null, namesValue: null }) {
    const catalogData: ShopCatalog = {};
    const dataGoods = goodsValue ?? data.Value.Goods;
    const dataNames = namesValue ?? names;
    const currencyValue = rootGetters['user/getUserCurrencyValue'];

    for (const good of dataGoods) {
      const currentGroup = dataNames[good.G];
      const currentGood = currentGroup.B[good.T];

      const normalizedGroup : ShopCatalogGroup = {
        id: good.G,
        items: [],
        ...currentGroup
      };

      const normalizedGood : ShopCatalogItem = {
        id: good.T,
        price: Math.round(good.C * currencyValue * 100) / 100,
        count: good.P,
        ...currentGood
      };

      const savedGroup = catalogData[normalizedGroup.id];

      if ( !savedGroup ) {
        normalizedGroup.items.push(normalizedGood);

        catalogData[normalizedGroup.id] = normalizedGroup;
      } else {
        savedGroup.items.push(normalizedGood);
      }
    }

    const sortCatalogData = Object.keys(catalogData).reduce((accum: Object[], currentValue) => {
      accum.push(catalogData[currentValue]);

      return accum;
    }, []);

    commit('SET_SHOP_CATALOG', sortCatalogData);

    return sortCatalogData;
  },

  fetchCatalogWithCurrency ({ commit, dispatch }: ActionContext<ShopCatalogState, any>) {
    const goodsData = require('../../static/data/data.json');
    const catalogParams = {
      goodsValue: goodsData.Value.Goods,
      namesValue: require('../../static/data/names.json')
    };
    const newCurrencyValue = getRandomNumberBetween(20, 81);

    commit('user/SET_CURRENCY_VALUE', newCurrencyValue, { root: true });

    return dispatch('getShopCatalog', catalogParams).then((response: ShopCatalogClient) => {
      return dispatch('user/updateCartPrices', response, { root: true });
    });
  }
};

export default actions;
