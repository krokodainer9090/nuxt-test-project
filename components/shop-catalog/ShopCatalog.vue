<script>
// Components
import ShopCatalogGroup from './shop-catalog-group/ShopCatalogGroup';

// Vuex
import { createNamespacedHelpers } from 'vuex';

// Store
const shopCatalogStore = createNamespacedHelpers( 'shop-catalog' );

export default {
  name: 'ShopCatalog',

  components: {
    ShopCatalogGroup
  },

  computed: {
    ...shopCatalogStore.mapGetters({
      shopCatalog: 'getShopCatalog'
    })
  },

  methods: {
    getShopCatalogGroupKey (key) {
      return `shop-catalog-section-${key}`;
    },

    onClickItem (item) {
      this.$store.dispatch('user/addItemToCart', item);
    }
  }
};
</script>

<template>
  <section class="shop-catalog w-100">
    <div class="container">
      <h2 class="shop-catalog__title">Каталог магазина</h2>
      <div class="row">
        <ShopCatalogGroup
          v-for="(group, index) in shopCatalog"
          :key="getShopCatalogGroupKey(group.id || index)"
          :group="group"
          class="shop-catalog__group col-12 col-md-6"
          @on-click-item="onClickItem"
        />
      </div>
    </div>
  </section>
</template>

<style lang="scss">
  @import "../../styles/components/shop-catalog/shop-catalog";
</style>
