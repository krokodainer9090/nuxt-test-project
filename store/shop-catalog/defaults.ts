import { ShopCatalogClient } from '~/store/shop-catalog/types';

export const defaultShopCatalog: ShopCatalogClient = [];
