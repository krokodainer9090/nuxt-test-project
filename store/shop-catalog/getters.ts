import { ShopCatalogState, ShopCatalogClient } from '~/store/shop-catalog/types';

export default {
  getShopCatalog (state: ShopCatalogState): ShopCatalogClient {
    return state.shopCatalog;
  }
};
