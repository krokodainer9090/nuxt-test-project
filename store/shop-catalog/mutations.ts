// Types
import { ShopCatalogClient, ShopCatalogState } from '~/store/shop-catalog/types';

// Vuex
import { MutationTree } from 'vuex';

const mutations: MutationTree<ShopCatalogState> = {
  SET_SHOP_CATALOG (state: ShopCatalogState, shopCatalog: ShopCatalogClient) {
    state.shopCatalog = shopCatalog;
  }
};

export default mutations;
