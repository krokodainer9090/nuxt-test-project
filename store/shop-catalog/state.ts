// Default
import { defaultShopCatalog } from '~/store/shop-catalog/defaults';

// Types
import { ShopCatalogState } from '~/store/shop-catalog/types';

export default (): ShopCatalogState => {
  return {
    shopCatalog: defaultShopCatalog
  };
};
