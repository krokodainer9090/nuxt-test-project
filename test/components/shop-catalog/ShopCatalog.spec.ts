import Vuex from 'vuex';

// Components
import ShopCatalog from '~/components/shop-catalog/ShopCatalog.vue';
import ShopCatalogGroup from '~/components/shop-catalog/shop-catalog-group/ShopCatalogGroup.vue';

// Utilities
import { shallowMount, Wrapper, createLocalVue } from '@vue/test-utils';

// Store
import { shopCatalogModule } from '~/test/store/shop-catalog/shop-catalog-store';
import { userModule } from '~/test/store/user/user-store';

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store({
  modules: {
    ...shopCatalogModule,
    ...userModule
  }
});

describe('ShopCatalog', () => {
  const mountFunction: (options?: Object) => Wrapper<any> = (options = {}) => {
    return shallowMount(ShopCatalog, {
      ...options,
      localVue,
      store
    });
  };

  it('should render empty catalog', () => {
    const wrapper = mountFunction();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render catalog with groups', async () => {
    // Call actions to get shop catalog
    await store.dispatch('shop-catalog/getShopCatalog' );

    // Check snapshot
    const wrapper = mountFunction();
    expect(wrapper.html()).toMatchSnapshot();

    wrapper.findComponent(ShopCatalogGroup).vm.$emit('on-click-item');
  });
});
