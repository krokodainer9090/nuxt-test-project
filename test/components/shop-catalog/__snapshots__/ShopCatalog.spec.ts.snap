// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopCatalog should render catalog with groups 1`] = `
"<section class=\\"shop-catalog w-100\\">
  <div class=\\"container\\">
    <h2 class=\\"shop-catalog__title\\">Каталог магазина</h2>
    <div class=\\"row\\">
      <shopcataloggroup-stub group=\\"[object Object]\\" class=\\"shop-catalog__group col-12 col-md-6\\"></shopcataloggroup-stub>
      <shopcataloggroup-stub group=\\"[object Object]\\" class=\\"shop-catalog__group col-12 col-md-6\\"></shopcataloggroup-stub>
      <shopcataloggroup-stub group=\\"[object Object]\\" class=\\"shop-catalog__group col-12 col-md-6\\"></shopcataloggroup-stub>
      <shopcataloggroup-stub group=\\"[object Object]\\" class=\\"shop-catalog__group col-12 col-md-6\\"></shopcataloggroup-stub>
      <shopcataloggroup-stub group=\\"[object Object]\\" class=\\"shop-catalog__group col-12 col-md-6\\"></shopcataloggroup-stub>
    </div>
  </div>
</section>"
`;

exports[`ShopCatalog should render empty catalog 1`] = `
"<section class=\\"shop-catalog w-100\\">
  <div class=\\"container\\">
    <h2 class=\\"shop-catalog__title\\">Каталог магазина</h2>
    <div class=\\"row\\"></div>
  </div>
</section>"
`;
